consensus/core/src/core.rs

```diff
@@ -749,6 +829,190 @@ impl Core {
         ancestors
     }
 
+    /// Retrieves the next ancestors to propose to form a block at `clock_round` round.
+    /// If smart selection is enabled then this will try to select the best ancestors
+    /// based on the propagation scores of the authorities.
+    fn smart_ancestors_to_propose(
+        &mut self,
+        clock_round: Round,
+        smart_select: bool,
+    ) -> Vec<VerifiedBlock> {
```

Comment:
If not as part of this PR, but I would strongly suggest maybe on a follow up PR adding more tests for this method. I believe it does quite a bit of work and it would be good to cover the various scenarios. I see below that you cover the case of the low score node, but maybe we want to test the other branches as well.

Reply:
Will do! Cases that still need to be tested in unit tests are
- Case when we attempt to propose ancestors but we do not have enough ancestors in INCLUDE state and so we wait for more ancestors.
- Case when we need to exclude EXCLUDE ancestors from parent round because we have not reached parent round quorum threshold
- Case when we need to include a weak link ancestor from dag state between last included round and network low quourm round.

Reply:
Nice, yes all those scenarios are valuable to test :)

> Case when we need to exclude EXCLUDE ancestors from parent round because we have not reached parent round quorum threshold

I assume you mean "include EXCLUDE ancestors" right?

Reply:
Yeah, updated the comment

Reply:
Added tests